Core/HW/StereoResampler.cpp

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) 2015- PPSSPP Project and Dolphin Project.
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 2.0 or later versions.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License 2.0 for more details.
+
+// A copy of the GPL 2.0 should have been included with the program.
+// If not, see http://www.gnu.org/licenses/
+
+// Official git repository and contact information can be found at
+// https://github.com/hrydgard/ppsspp and http://www.ppsspp.org/.
+
+// Adapted from Dolphin.
+
+#include <string.h>
+
+#include "base/logging.h"
+#include "Common/ChunkFile.h"
+#include "Common/MathUtil.h"
+#include "Common/Atomics.h"
+#include "Core/HW/StereoResampler.h"
+#include "Globals.h"
+
+#ifdef _M_SSE
+#include <emmintrin.h>
+#endif
+
+inline void ClampBufferToS16(s16 *out, const s32 *in, size_t size) {
+#ifdef _M_SSE
+	// Size will always be 16-byte aligned as the hwBlockSize is.
+	while (size >= 8) {
+		__m128i in1 = _mm_loadu_si128((__m128i *)in);
+		__m128i in2 = _mm_loadu_si128((__m128i *)(in + 4));
+		__m128i packed = _mm_packs_epi32(in1, in2);
+		_mm_storeu_si128((__m128i *)out, packed);
+		out += 8;
+		in += 8;
+		size -= 8;
+	}
+	for (size_t i = 0; i < size; i++) {
+		out[i] = clamp_s16(in[i]);
+	}
+#else
+	for (size_t i = 0; i < size; i++) {
+		out[i] = clamp_s16(in[i]);
+	}
+#endif
+}
+
+void StereoResampler::MixerFifo::Clear() {
+	memset(m_buffer, 0, sizeof(m_buffer));
+	// TODO
```

Comment:
What's left todo other than clearing it?  Do we need to tweak the indexes so that it doesn't think there are valid 0s?

-[Unknown]